### DISCUSSION TITLE
Max session size

### DISCUSSION THREAD

**Discussion by brentryan** [awslabs/aws-dotnet-distributed-cache-provider#41] [Q&A] [answered]
What is the max size of a session that I can have? Is it limited by the dynamodb item size limit of 400KB?

**Comment by normj** [comment, marked as answer]
Yes, the size is limited to DynamoDB's item size. Do you expect to need a larger size?

**Reply by brentryan** [reply]
Yes. Unfortunately we have legacy software that relies on some use cases that have larger than 1MB sessions. The dynamodb limit pushes us towards couchbase/elasticache instead which has some other performance benefits but it's not as cost effective for our workloads. 

The fix for us is to either replace all the code with a different design that doesn't rely on large session state payloads (mountains of code so not small effort) or keep using elaticache and eat the costs.

**Reply by ashovlin** [reply]
We may be able to explore an option to compress the session data we get from ASP.NET before storing it in DynamoDB. 

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/bp-use-s3-too.html#bp-use-s3-too-or-compress

**Comment by normj** [comment]
In theory we could break the session over multiple DDB items and use a range key. That would add significant complexity to the library managing the session of multiple items. I would suggest creating a feature request issue asking for support for cache size larger then a DynamoDB. If we get enough interest in the issue we can prioritizing adding the support.

**Reply by rajeshaz09** [reply]
1 year back i implemented dynamodb distributed cache for asp.net core in one of the project. We faced similar issue. I followed below logic

Use [messagepack for c#](https://github.com/MessagePack-CSharp/MessagePack-CSharp) to compress data in efficient way

Split message into multiple based on max dynamodb size then introduced sort key as message number

When reading... get all items with partition key then merge data as per sort key 

Below optional optimisation i did

During writing make sure you delete items with sortkey greater than message split count. In rare cases we may face this problem. Initially there might be N split messages, but when you update cache, you may set value to much lesser data which may endup less than N messages.
